optaplanner-core/src/test/java/org/optaplanner/core/api/score/stream/quad/QuadConstraintStreamTest.java

```diff
@@ -0,0 +1,311 @@
+    }
+
+    @Test
+    public void reward_Long() {
+        assumeDrools();
+        TestdataSimpleLongScoreSolution solution = new TestdataSimpleLongScoreSolution();
+        TestdataValue v1 = new TestdataValue("v1");
+        solution.setValueList(Arrays.asList(v1));
+        TestdataEntity e1 = new TestdataEntity("e1", v1);
+        TestdataEntity e2 = new TestdataEntity("e2", v1);
+        solution.setEntityList(Arrays.asList(e1, e2));
+
+        ConstraintStreamScoreDirectorFactory<TestdataSimpleLongScoreSolution> scoreDirectorFactory
+                = new ConstraintStreamScoreDirectorFactory<>(
+                TestdataSimpleLongScoreSolution.buildSolutionDescriptor(), (factory) -> new Constraint[]{
+                factory.from(TestdataEntity.class)
+                        .join(TestdataValue.class, equal(TestdataEntity::getValue, identity()))
+                        .join(TestdataValue.class)
+                        .reward("myConstraint1", SimpleLongScore.ONE),
+                factory.from(TestdataEntity.class)
+                        .join(TestdataValue.class, equal(TestdataEntity::getValue, identity()))
+                        .join(TestdataValue.class)
+                        .rewardLong("myConstraint2", SimpleLongScore.ONE, (entity, value, extra) -> 20L)
+        }, constraintStreamImplType);
+        InnerScoreDirector<TestdataSimpleLongScoreSolution> scoreDirector = scoreDirectorFactory.buildScoreDirector(false, constraintMatchEnabled);
+
+        scoreDirector.setWorkingSolution(solution);
+        assertEquals(SimpleLongScore.of(42L), scoreDirector.calculateScore());
+    }
+
+    @Test
+    public void reward_BigDecimal() {
+        assumeDrools();
+        TestdataSimpleBigDecimalScoreSolution solution = new TestdataSimpleBigDecimalScoreSolution();
+        TestdataValue v1 = new TestdataValue("v1");
+        solution.setValueList(Arrays.asList(v1));
+        TestdataEntity e1 = new TestdataEntity("e1", v1);
+        TestdataEntity e2 = new TestdataEntity("e2", v1);
+        solution.setEntityList(Arrays.asList(e1, e2));
+
+        ConstraintStreamScoreDirectorFactory<TestdataSimpleBigDecimalScoreSolution> scoreDirectorFactory
+                = new ConstraintStreamScoreDirectorFactory<>(
+                TestdataSimpleBigDecimalScoreSolution.buildSolutionDescriptor(), (factory) -> new Constraint[]{
+                factory.from(TestdataEntity.class)
+                        .join(TestdataValue.class, equal(TestdataEntity::getValue, identity()))
+                        .join(TestdataValue.class)
+                        .reward("myConstraint1", SimpleBigDecimalScore.ONE),
+                factory.from(TestdataEntity.class)
+                        .join(TestdataValue.class, equal(TestdataEntity::getValue, identity()))
+                        .join(TestdataValue.class)
+                        .rewardBigDecimal("myConstraint2", SimpleBigDecimalScore.ONE, (entity, value, extra) -> new BigDecimal("0.2"))
```

Comment:
This looks to me as if it creates a Tri stream, not a quad stream.

Reply:
the matcher confirms that, clearly a tri stream matcher.

Reply:
Other cases above too - the quad streams tests is really good at testing tri streams :)